crates/next-dev/src/lib.rs

```diff
@@ -143,29 +150,75 @@ impl NextDevServerBuilder {
         let console_ui = Arc::new(ConsoleUi::new(log_options));
         let console_ui_to_dev_server = console_ui.clone();
 
-        let server = DevServer::listen(
-            turbo_tasks.clone(),
-            move || {
-                source(
-                    root_dir.clone(),
-                    project_dir.clone(),
-                    entry_requests.clone(),
-                    eager_compile,
-                    turbo_tasks.clone().into(),
-                    console_ui.clone().into(),
-                    browserslist_query.clone(),
-                    server_component_externals.clone(),
-                )
-            },
-            (
-                self.hostname.context("hostname must be set")?,
-                self.port.context("port must be set")?,
+        let start_port = self.port.context("port must be set")?;
+        let host = self.hostname.context("hostname must be set")?;
+
+        let mut err: Option<anyhow::Error> = None;
+
+        let tasks = turbo_tasks.clone();
+        let source = move || {
+            source(
+                root_dir.clone(),
+                project_dir.clone(),
+                entry_requests.clone(),
+                eager_compile,
+                turbo_tasks.clone().into(),
+                console_ui.clone().into(),
+                browserslist_query.clone(),
+                server_component_externals.clone(),
             )
-                .into(),
-            console_ui_to_dev_server,
-        );
+        };
+
+        // Retry to listen on the different port if the port is already in use.
+        for retry_count in 0..10 {
```

Comment:
Nit: I think this might be in the wrong location, and it's because we coupled `build`ing a server instance with connecting the server to a port. Why not return a `DevServer` instance and allow the caller to invoke `listen` with their own incrementing logic?

Reply:
Could be, but since there's a single caller currently (start_server) and I wanted to preserve the existing flow for now. As you said `NextDevServerBuilder` does something encapsulated manner by bringing all options and try to listen / bind itself internally - refactoring to extract out felt to be better as different PR to consolidate.